parser.py

```diff
@@ -1043,18 +1093,46 @@ def main():
         shutil.copy('assets/images/favicon.ico', paths.file_tweet_icon)
 
     media_sources = parse_tweets(username, users, html_template, paths)
-    parse_followings(users, URL_template_user_id, paths)
-    parse_followers(users, URL_template_user_id, paths)
-    parse_direct_messages(username, users, URL_template_user_id, paths)
 
-    # find user ids to look up from group dms
+    following_ids = collect_user_ids_from_followings(paths)
+    print(f'found {len(following_ids)} user IDs in followings.')
+    follower_ids = collect_user_ids_from_followers(paths)
+    print(f'found {len(follower_ids)} user IDs in followers.')
+    dms_user_ids = collect_user_ids_from_direct_messages(paths)
+    print(f'found {len(dms_user_ids)} user IDs in direct messages.')
     group_dms_user_ids = collect_user_ids_from_group_direct_messages(paths)
-    # TODO: separate the collecting of user ids out of the other parse* functions in the same way
-    #  and pool the lookups together before all of the other parsing & output generation
-    # look them up
-    lookup_users(group_dms_user_ids, users)
+    print(f'found {len(group_dms_user_ids)} user IDs in group direct messages.')
+
+    # bulk lookup for user handles from followers, followings, direct messages and group direct messages
+    collected_user_ids_without_followers = list(
+        set(following_ids).union(set(dms_user_ids)).union(set(group_dms_user_ids))
+    )
+    collected_user_ids_only_in_followers: set = set(follower_ids).difference(set(collected_user_ids_without_followers))
+    collected_user_ids: list = list(set(collected_user_ids_without_followers).union(collected_user_ids_only_in_followers))
+
+    print(f'\nfound {len(collected_user_ids)} user IDs overall.')
+
+    # give the user a choice if followers should be included in the lookup
+    # (but only in case they make up a large amount):
+    unknown_collected_user_ids: set = set(collected_user_ids).difference(users.keys())
+    if len(unknown_collected_user_ids) > 10000:
+        unknown_follower_user_ids: set = unknown_collected_user_ids.intersection(collected_user_ids_only_in_followers)
+        if len(unknown_follower_user_ids) > 5000:
+            # Account metadata observed at ~2.1KB on average.
+            estimated_follower_lookup_size = int(2.1 * len(unknown_follower_user_ids))
+            user_input = input(f'{len(unknown_follower_user_ids)} of the {len(unknown_collected_user_ids)} '
```

Comment:
Currently, this gives an extra prompt if there are more than 10 000 unknown user handles, of which more than 5 000 are only from followers. I'm not sure if these values are sensible, but they could easily be changed. In the minimum case, this would save (up to) about 1.7 minutes of download time on a slow connection (max 2 seconds per 100 users, so max 100 seconds for 5000 users). What numbers would you suggest here?

Reply:
Sounds very reasonable.

Reply:
I simplified the check a bit more (don't check for > 10 000 in general, only for > 5000 from followers) and added a bit more clarification to the text of the prompt.

The approval was already dismissed anyway when I merged some other changes from main into this. So you might want to have a look again? Sorry for a lot of back and forth here...